README.md
# DAPNETAPRS
APRS Interface to display transmitters on APRS MAP

__Tasks:__
* Get transmitter status and location via curl request to DAPNET Core
* Open APRS IS connection
* Generate APRS objects with name PS-<CALLSIGN_FROM_DAPNET> (reduce to 7 letters if needed)
* Online transmitters: Active object with ``;``

Example: ``DH3WR-3>APRS,TCPIP*:;PS-DB0SDA*240909z5000.60NP00600.60E#PHG2230Testing DH3WR``
* Offline transmitters: Delete object from map with ``_``

Example: ``DH3WR-3>APRS,TCPIP*:;PS-DB0SDA*240909z5000.60NP00600.60E#PHG2230Testing DH3WR``

* Generate PHG comment according to curl transmitter data (Infos: http://www.tapr.org/pipermail/aprssig/2006-October/016629.html http://aprsisce.wikidot.com/phg

* Add owner in comment text from curl answer

Symbol is:
* Overlay for online transmitter ``P``
* Overlay for offline transmitter ``X``
* Symbol ``#``


CHANGELOG
June 4th. 2021
- changed logging function
- rewrite / streamline code
- option to select own aprs server
- reduced procedure to online transmiters only
- changed APRS comment field to [default comment][transmitter type][time slots]
- moved config to /etc
- simplified check with mask table

May 5th. 2017
- restyled configuration file to be read via configparser
- switched to urllib3, redesigned http request procedure
- removed indent error in PHG conversion
- extended DEPENDENCIES
- added CHANGELOG
- edited INSTALL


config.py
#configuration file for DAPNETAPRS
#store this file in a safe directory, out of range of any kind of server

user = dict(
	username = '', #your DAPNET username
	password = '', #your DAPNET password
	datenotation = 'nl' #date notation (for weather messages) (en/de/nl)
)

dapnet = dict(
	callsign = '', #target callsign for weather messages
	transmittergrp = '', #transmittergroup for weather messages
	baseurl = 'http://www.hampager.de:8080', #base url of DAPNET API
	coreurl = '/calls', #path to send calls
	trxurl = '/transmitters' #path to query transmitters
)

aprsis = dict(
	username = '', #APRS-IS callsign
	password = '', #APRS-IS pass code
	sourcecall = '', #source call when placing objects
	apikey = '' #APRS-IS API key
)

misc = dict(
	logfile = '/var/tmp/dapnetaprs.log' #log file with full path
)


testing.py
#!/usr/bin/env python
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Author: Ralf Wilke DH3WR  dh3wr atREMOVETHIS darc .REMOVETHIS de
# Date: 25.04.2017
# Version 0.1

import aprslib
import logging
import time
from time import sleep
from datetime import datetime
from aprslib.packets.base import APRSPacket
from aprslib.util import latitude_to_ddm, longitude_to_ddm, comment_altitude
import urllib2
import json
import base64
import math

logging.basicConfig(level=logging.DEBUG) # level=10

hampagerusername = 'dh3wr'
hampagerpassword = ''
hampagerurl = "http://www.hampager.de:8080/transmitters"

aprsisusername = 'DH3WR'
aprsispassword = ''
aprsissourcecallsign = 'DB0SDA-14'


def encodePHG ( power, height, gain, dir ):
	p = round(math.sqrt(power))
	if p > 9:
		p = 9

        if height <= 0:
		h = 0
	else:
	        h = round(math.log(height/10/0.3048,2))
		if h > 9:
			h = 9	
	g = round(gain)
	if g > 9:
		g = 9

	if dir <= -1:
		d = 0
	else:
		d = round(dir/45)
	if d > 9:
		d = 9

	return ('PHG' + "{:.0f}".format(p) + "{:.0f}".format(h) + "{:.0f}".format(g) + "{:.0f}".format(d))


AIS = aprslib.IS(aprsisusername, passwd=aprsispassword, port=14580)
AIS.connect()


request = urllib2.Request(hampagerurl)
base64string = base64.b64encode('%s:%s' % (hampagerusername, hampagerpassword))
request.add_header("Authorization", "Basic %s" % base64string)
response = urllib2.urlopen(request)
dapnetdata = json.loads(response.read())
#print (dapnetdata[0])


for tx in range (0, len(dapnetdata)):
	mytx = dapnetdata[tx]
	longitude = float(mytx['longitude'])
	latitude = float(mytx['latitude'])
	format = 'uncompressed'
	if mytx['status'] == 'ONLINE':
		active = '*'
	else:
		active = '_'

	callsign = mytx['name'].upper()
	
	# Crop callsign to 6 characters
	if len(callsign) > 6:
		callsign = callsign[:6]
	callsign = "{:<6}".format(callsign)
	timeSlots = mytx['timeSlot']
	symbol_table = 'P'
	symbol = '#'
	altitude = None
	comment = 'DAPNET POCSAG Transmitter'
	timestamp = datetime.utcfromtimestamp(time.time()).strftime("%d%H%M") + 'z'

	if mytx['antennaType'] == 'OMNI':
		direction = -1
	else:
		direction = mytx['antennaDirection']

	phg = encodePHG(float(mytx['power']),float(mytx['antennaAboveGroundLevel']),float(mytx['antennaGainDbi']),direction)
	if active == '_':
		pgh = ''
	body = [
		aprsissourcecallsign,
		'>APRS,TCPIP*:',
		';',
		'PS-',
		callsign,
		active,
		timestamp,
		latitude_to_ddm(latitude),
		symbol_table,
		longitude_to_ddm(longitude),
		symbol,
		phg,
		comment,
		' Timeslots: ',
		timeSlots,
	]

	data = "".join(body)
	print(data)

	AIS.sendall(data)
	sleep(0.3)


newdapaprs.py
import aprslib
import configparser
import urllib3
import json
import sys
import math
import time
from time import sleep
from datetime import datetime
import MySQLdb
from logfunctions import *
from aprslib.util import latitude_to_ddm, longitude_to_ddm, comment_altitude

# In case callsigns aren't published these will be collected in this list
errorcalls = []

# read configuration
cfg = configparser.ConfigParser()
try:
	cfg.read('/etc/dapaprs.cfg')
	dapnet_uri = cfg.get('dapnet','baseurl') + cfg.get('dapnet','trxurl')
	mask_server = cfg.get('maskdb','server')
	mask_database = cfg.get('maskdb','database')
	mask_username = cfg.get('maskdb','username')
	mask_password = cfg.get('maskdb','passwd')
	aprsissourcecallsign = cfg.get('aprsis','sourcecall')
	aprsisuser = cfg.get('aprsis','username')
	aprsispasswd = cfg.get('aprsis','password')
	aprsishost = cfg.get('aprsis','host')
	aprsisport = int(cfg.get('aprsis','port'))
	aprsissleep = float(cfg.get('aprsis','sleeptime'))
except:
	logger.error('Cannot read config file')
	sys.exit(0)

# setup aprs is link
def APRSConnect(aprsisuser,aprsispassword,aprstelegram):
	try:
		ais = aprslib.IS(aprsisuser,passwd = aprsispasswd, host = aprsishost ,port = aprsisport)
		ais.connect()
		ais.sendall(aprstelegram)
	except:
		aprs_logger.error('Can not connect to APRS-IS')
		sys.exit(0)

# read dapnet transmitters
def GetTXData():
	try:
		http = urllib3.PoolManager()
		trxdata = http.request('GET',dapnet_uri)
		trxdata = json.loads(trxdata.data.decode('utf-8'))
		return(trxdata)
	except:
		logger.error('Cannot retrieve DAPNET transmitter data')
		sys.exit(0)

# read data from masking table
# on fail script will be quited
def GetMaskData():
	try:
		db = MySQLdb.connect(host=mask_server,user=mask_username,passwd=mask_password,db = mask_database)
		cur = db.cursor()
		cur.execute('SELECT MASK_CALL FROM masktable')
		mask_result = cur.fetchall()
		mask_list = []
		for row in mask_result:
			mask_list.append(row[0])
		return(mask_list)
	except:
		logger.error('Cannot connect to mask database')
		sys.exit(1)

# Determine whether or not callsign will be published as object
def IsTXMasked(callsign):
	if callsign in mask_data:
		return(True)
	else:
		return(False)

# Encode PHG data using
def EncodePHG(power, height, gain, dir):
	p = round(math.sqrt(power))
	if p > 9:
		p = 9
	if height <= 0:
		h = 0
	else:
		h = round(math.log(height/10/0.3048,2))
	if h <= 0:
		h = 0
	if h > 9:
		h = 9
	g = round(gain)
	if g > 9:
		g = 9
	if dir <= -1:
		d = 0
	else:
		d = round(dir / 45)
	if d > 9:
		d = 9
	return ('PHG' + "{:.0f}".format(p) + "{:.0f}".format(h) + "{:.0f}".format(g) + "{:.0f}".format(d))

# Main loop through transmitter data as received from DAPNET API
def LoopTx(trxdata,maskdata):
	for tx in trxdata:
		callsign = tx['name'][:7].upper()
		callsign = "{:<7}".format(callsign)
		if IsTXMasked(callsign) == False and tx['status'] == 'ONLINE':
			phg = (EncodePHG(float(tx['power']),float(tx['antennaAboveGroundLevel']),float(tx['antennaGainDbi']),float(tx['antennaDirection'])))
			latitude = latitude_to_ddm(float(tx['latitude']))
			longitude = longitude_to_ddm(float(tx['longitude']))
			timeslots = ' (' + tx['timeSlot'] + ')'
			txtype = tx['deviceType']
			timestamp = datetime.utcfromtimestamp(time.time()).strftime("%d%H%M") + 'z'

			# fix fields
			prefix = '>APRS,TCPIP*:;P-'
			activetx = '*'
			symbol = '#'
			symboltable = 'P'
			comment = 'DAPNET POCSAG TX'

			# compose message body
			body = [
				aprsissourcecallsign,
				prefix,
				callsign,
				activetx,
				timestamp,
				latitude,
				symboltable,
				longitude,
				symbol,
				phg,
				comment,
				': ',
				txtype,
				timeslots,
			]
			aprstelegram = ''.join(body)
			try:
				APRSConnect(aprsisuser,aprsispasswd,aprstelegram)
				sleep(aprsissleep)
			except:
				errorcalls.append(callsign)
				aprs_logger.error('Can not send APRS telegram (' + aprstelegram + ')')
logger.debug('Start loop')
trxdata = GetTXData()
mask_data = GetMaskData()
LoopTx(trxdata,mask_data)
if len(errorcalls) > 0:
	aprs_logger.error(errorcalls)
logger.debug('Loop completed')


dapnetaprs.py
#!/usr/bin/env python3
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Author: Ralf Wilke DH3WR  dh3wr atREMOVETHIS darc .REMOVETHIS de
# Author: Ronald Bouwens PE2KMV pe2kmv atREMOVETHIS gmail .REMOVETHISPART c0m
# Date: 25.11.2018
# Version 1.1

import aprslib
import logging
import time
from time import sleep
from datetime import datetime
from aprslib.packets.base import APRSPacket
from aprslib.util import latitude_to_ddm, longitude_to_ddm, comment_altitude
import urllib3
import json
import base64
import math
import sys
import configparser
import os
import MySQLdb

logging.basicConfig(filename='dapnet.log',level=logging.CRITICAL)

#assign configuration file
cfg = configparser.RawConfigParser()
try:
	#attempt to read the config file config.cfg
	config_file = os.path.join(os.path.dirname(__file__),'config.cfg')
	cfg.read(config_file)
except:
	#no luck reading the config file, write error and bail out
	logger.error('APRS script could not find / read config file')
	sys.exit(0)


#setup the logging system. Error logs are written in logfile as determined by config
logger = logging.getLogger('dapnet')
handler = logging.FileHandler(cfg.get('misc','logfile'))
logformat = logging.Formatter('%(asctime)s %(levelname)s %(message)s')
handler.setFormatter(logformat)
logger.addHandler(handler)
logger.setLevel(logging.WARNING)

#DAPNET credentials and API URL are read from config file
hampagerusername = cfg.get('user','username')
hampagerpassword = cfg.get('user','password')
hampagerurl = cfg.get('dapnet','baseurl') + cfg.get('dapnet','trxurl')

#APRS-IS credentials and settings are read from the config file
aprsisusername = cfg.get('aprsis','username')
aprsispassword = cfg.get('aprsis','password')
aprsissourcecallsign = cfg.get('aprsis','sourcecall')

#read settings for maskind database
maskserver = cfg.get('maskdb','server')
maskuser = cfg.get('maskdb','username')
maskpasswd = cfg.get('maskdb','passwd')
maskdatabase = cfg.get('maskdb','database')

#open database connection
db = MySQLdb.connect(host=maskserver,user=maskuser,passwd=maskpasswd,db=maskdatabase)

if not (db):
	#bail out if database connection fails
	logger.error('No connecting to masking database')
	sys.exit()

#read calls from masking database and table
cur = db.cursor()
cur.execute("SELECT MASK_CALL FROM masktable")
excl_result = cur.fetchall()
excl_list = []
for row in excl_result:
	excl_list.append(row[0])

#create a PHG code using TRX power, antenna height and antenna gain from DAPNET data
def encodePHG (power, height, gain, dir):
	p = round(math.sqrt(power))
	if p > 9:
		p = 9
	if height <= 0:
		h = 0
	else:
	    h = round(math.log(height/10/0.3048,2))
	if h > 9:
		h = 9	
	g = round(gain)
	if g > 9:
		g = 9
	if dir <= -1:
		d = 0
	else:
		d = round(dir/45)
	if d > 9:
		d = 9
	return ('PHG' + "{:.0f}".format(p) + "{:.0f}".format(h) + "{:.0f}".format(g) + "{:.0f}".format(d))


#open the link to APRS-IS. If connnection can't be established, print warning to console, print warning to error log and bail out
try:
	AIS = aprslib.IS(aprsisusername, passwd=aprsispassword, port=14580)
	AIS.connect()
except:
	print('Invalid APRS credentials')
	logger.error('Invalid APRS credentials')
	sys.exit(0)
else:
	#connection to APRS-IS has been established, now continue
	
	#create the complete URL to send to DAPNET
	http = urllib3.PoolManager()
	headers = urllib3.util.make_headers(basic_auth= hampagerusername + ':' + hampagerpassword)

try:
	#try to establish connection to DAPNET
	response = http.request('GET',hampagerurl,headers=headers)
except:
	#connection to DAPNET failed, write warning to console, write warning to error log then bail out
	print('Invalid DAPNET credentials')
	logger.error('Invalid DAPNET credentials')
	sys.exit(0)
else:
	#connection to DAPNET has been established, continue
	dapnetdata = json.loads(response.data.decode('utf-8'))

#loop through all transmitters returned in dapnetdata variable
for tx in range (0, len(dapnetdata)):
	mytx = dapnetdata[tx] #get the station call
	longitude = float(mytx['longitude']) #station longitude
	latitude = float(mytx['latitude']) #station latitude
	format = 'uncompressed' #aprs coordinates format
	if mytx['status'] == 'ONLINE': #determine whether DAPNET returns status ONLINE / OFFLINE
		#pager transmitter is online, add in 'create object' symbol
		active = '*'
		symbol = '#' #green digi star - transmitter online
	else:
		#pager transmitter is offline, add in 'kill object' symbol
		#note that aprs.fi doesn't support killing objects!
		active = '_'
		symbol = 'n' #red triangle - transmitter offline

	callsign = mytx['name'].upper() #station call is converted into upper case
	
	# Crop callsign to 6 characters
	if len(callsign) > 6:
		callsign = callsign[:6]
	callsign = "{:<6}".format(callsign)
	symbol_table = 'P' #add in APRS symbol character
#	symbol = '#' #add in APRS symbol
	altitude = None #don't use altitude
	comment = 'DAPNET POCSAG Transmitter(' #add comment to object
	timeSlots = mytx['timeSlot'] + ')' #extract timeslots for station
	status = '(' + mytx['status'] + ')'
	timestamp = datetime.utcfromtimestamp(time.time()).strftime("%d%H%M") + 'z'
	#differentiate between omnidirectonal antenna or beam
	if mytx['antennaType'] == 'OMNI':
		direction = -1 #add in string to mark 'omnidirectional'
	else:
		direction = mytx['antennaDirection'] #add in beaming direction if applicable

	#jump back and calculate PHG code with DAPNET data as input
	phg = encodePHG(float(mytx['power']),float(mytx['antennaAboveGroundLevel']),float(mytx['antennaGainDbi']),direction)
	if active == '_':
		pgh = ''
	#put all variables in the correct order in a list
	body = [
		aprsissourcecallsign,
		'>APRS,TCPIP*:',
		';',
		'PS-',
		callsign,
		active,
		timestamp,
		latitude_to_ddm(latitude),
		symbol_table,
		longitude_to_ddm(longitude),
		symbol,
		phg,
		comment,
		timeSlots,
		status,
	]
	#create the complete message string out of all components
	data = "".join(body)
	#transmitters with status ERROR are ignored
	#edit March 29th. only show online transmitters
	if (mytx['status'] == "ONLINE") & (str.strip(callsign) not in excl_list):
		try:
			AIS.sendall(data)
#			print(data) #print the APRS message string to the console
			sleep(0.3)
		except:
			#APRS string could not be sent, log error, log string then continue
			logger.error('APRS message not sent to server')
			logger.info(data)

